garage/tf/regressors/categorical_mlp_regressor_with_model.py

```diff
@@ -0,0 +1,273 @@
+            consists of two hidden layers, each with 32 hidden units.
+        hidden_nonlinearity (callable): Activation function for intermediate
+            dense layer(s). It should return a tf.Tensor. Set it to
+            None to maintain a tanh activation.
+        hidden_w_init (callable): Initializer function for the weight
+            of intermediate dense layer(s). The function should return a
+            tf.Tensor. Default is Glorot uniform initializer.
+        hidden_b_init (callable): Initializer function for the bias
+            of intermediate dense layer(s). The function should return a
+            tf.Tensor. Default is zero initializer.
+        output_nonlinearity (callable): Activation function for output dense
+            layer. It should return a tf.Tensor. Set it to None to
+            maintain a softmax activation.
+        output_w_init (callable): Initializer function for the weight
+            of output dense layer(s). The function should return a
+            tf.Tensor. Default is Glorot uniform initializer.
+        output_b_init (callable): Initializer function for the bias
+            of output dense layer(s). The function should return a
+            tf.Tensor. Default is zero initializer.
+        optimizer (garage.tf.Optimizer): Optimizer for minimizing the negative
+            log-likelihood. Defaults to LbsgsOptimizer
+        optimizer_args (dict): Arguments for the optimizer. Default is None,
+            which means no arguments.
+        tr_optimizer (garage.tf.Optimizer): Optimizer for trust region
+            approximation. Defaults to ConjugateGradientOptimizer.
+        tr_optimizer_args (dict): Arguments for the trust region optimizer.
+            Default is None, which means no arguments.
+        use_trust_region (bool): Whether to use trust region constraint.
+        max_kl_step (float): KL divergence constraint for each iteration.
+        normalize_inputs (bool): Bool for normalizing inputs or not.
+        layer_normalization (bool): Bool for using layer normalization or not.
+    """
+
+    def __init__(self,
+                 input_shape,
+                 output_dim,
+                 name='CategoricalMLPRegressorWithModel',
+                 hidden_sizes=(32, 32),
+                 hidden_nonlinearity=tf.nn.tanh,
+                 hidden_w_init=tf.glorot_uniform_initializer(),
+                 hidden_b_init=tf.zeros_initializer(),
+                 output_nonlinearity=tf.nn.softmax,
+                 output_w_init=tf.glorot_uniform_initializer(),
+                 output_b_init=tf.zeros_initializer(),
+                 optimizer=None,
+                 optimizer_args=None,
+                 tr_optimizer=None,
+                 tr_optimizer_args=None,
+                 use_trust_region=True,
+                 max_kl_step=0.01,
+                 normalize_inputs=True,
+                 layer_normalization=False):
+
+        super().__init__(input_shape, output_dim, name)
```

Comment:
Maybe it will save us some time in the future, if we add a checking / assert here for input_shape.

Reply:
what check, specifically? i guess that it has 2 dimensions only?

Reply:
Shouldn't it be single dimension? Since the inputs are flattened, like `len(input_shape) == 1`?

Reply:
does it support a batch dimension?

Reply:
I think we usually do `shape=(None, ) + input_shape` like at line 123, so batch dimension is expected.

Reply:
Yes, it supports batch dimension (also tested in tests), restriction is just that individual data points should be one dimensional

Reply:
okay -- we can improve the interface for primitives once we get the models transition done.